### PR TITLE
use pydantic v2 for schema generation where possible

### PR DESCRIPTION
**Pull request by jakekaplan**
Previously we were doing something like the following when generating schema for flows/deployments:

```
if HAS_PYDANTIC_V2 and has_v2_type_as_param(signature):
    # build jsonschema with pydantic v2
else:
   # build jsonschema with pydantic v1
```       

This meant that unless the parameter passed was specifically a v2 pydantic model, we would default to v1. Notably this meant that all flow signatures that didn't include any pydantic models would use pydantic v1.

Generating schemas with pydantic v1 is frankly bad. Many of the issues with schema generation in pydantic v1 were only fixed in v2. To this end, until we can live in a fully v2 world, we want to at least try and use pydantic v2 as much as possible for schema generation (instead of as little as possible). 

To do this, this pr makes the following change:

```
if HAS_PYDANTIC_V2 and not has_v1_type_as_param(signature):
    # build jsonschema with pydantic v2
else:
   # build jsonschema with pydantic v1
```

### CODE REVIEW

File: tests/utilities/test_callables.py [hunk 1]
```diff
@@ -139,33 +139,63 @@ def f(
...
-                    "format": "date-time",
-                    "position": 0,
-                },
-                "y": {
-                    "title": "y",
-                    "default": "2025-01-01T00:00:00+00:00",
-                    "type": "string",
-                    "format": "date-time",
-                    "position": 1,
+        if HAS_PYDANTIC_V2:
```
Comment (jakekaplan) [inline tests/utilities/test_callables.py:142]: I opted to check this inside the test directly so it's easy to see the differences between the v1 and v2 versions of the schema

Comment (urimandujano) [reply tests/utilities/test_callables.py:142]: I like what you did with the schemas here!

**Review by urimandujano — COMMENTED** [review]
What do you think of adding a case to test that a function with a v1 object serializes to a v1 schema, otherwise we get a v2 schema? I think there might be a test case in here somewhere already and we might be able to swap the assert statement.

**Comment by jakekaplan** [comment]
> What do you think of adding a case to test that a function with a v1 object serializes to a v1 schema, otherwise we get a v2 schema? I think there might be a test case in here somewhere already and we might be able to swap the assert statement.

That's a good idea! let me take a look

File: src/prefect/_internal/pydantic/v2_schema.py [hunk 2]
```diff
@@ -106,7 +106,11 @@ def create_v2_schema(
     model = create_model(
         name_, __config__=model_cfg, __base__=model_base, **model_fields
     )
-    adapter = TypeAdapter(model)
+    try:
+        adapter = TypeAdapter(model)
+    except PydanticUndefinedAnnotation as exc:
+        # in v1 this raises a TypeError, which is handled by parameter_schema
```
Comment (jakekaplan) [inline src/prefect/_internal/pydantic/v2_schema.py:112]: In v1 `create_schema` raises a type error which is caught by `parameter_schema` to type to `Any`. I am not sure if this is the right way to handle this...

File: src/prefect/utilities/callables.py [hunk 3]
```diff
@@ -325,7 +325,7 @@ def parameter_schema(fn: Callable) -> ParameterSchema:
     class ModelConfig:
         arbitrary_types_allowed = True
 
-    if HAS_PYDANTIC_V2 and has_v2_type_as_param(signature):
+    if HAS_PYDANTIC_V2 and not has_v1_type_as_param(signature):
```
Comment (zzstoatzz) [inline src/prefect/utilities/callables.py:328]: 🔥

File: src/prefect/_internal/pydantic/v2_schema.py (hunk 2 above)
Comment (jakekaplan) [reply src/prefect/_internal/pydantic/v2_schema.py:112]: these tests were failing without it:
https://github.com/PrefectHQ/prefect/blob/main/tests/test_flows_compat.py#L1-L30

Comment (zzstoatzz) [reply src/prefect/_internal/pydantic/v2_schema.py:112]: hmm - I am also not totally sure about that 🧐

i don't immediately have a better suggestion though

**Review by zzstoatzz — APPROVED** [review]

**Review by zzstoatzz — CHANGES_REQUESTED** [review]
hmm actually I'm a little concerned by these failures
```python
AssertionError: expected call not found.
```

**Review by serinamarie — APPROVED** [review]
LGTM with the caveat that I don't know the answer to your question regarding handling of the `TypeError` in `TypeAdapter(model)`

**Comment by jakekaplan** [comment]
> hmm actually I'm a little concerned by [these](https://github.com/PrefectHQ/prefect/actions/runs/8380569156/job/22950084277?pr=12210#step:13:243) failures
> 
> ```python
> AssertionError: expected call not found.
> ```
> 
> @jakekaplan have you seen those before? I don't really see how they'd be so related this change though

I have not... I think there was an SSL change that got merged today. let me look into it

**Comment by jakekaplan** [comment]
@zzstoatzz I think the test failures were related to a different PR as best I can tell

**Review by zzstoatzz — APPROVED** [review]
lgtm

**Pull request merged**